cfgen/cfgen

```diff
@@ -476,9 +476,23 @@ def run_command(hostname: str, *args: str) -> str:
                                    timeout=15).decode()
 
 
+def validate_facter(hostname: str) -> None:
+    ver_str = run_command(hostname, 'facter', '--version')
+    # The output may contain commit id like this:
+    # 3.14.8 (commit 4339472f441868ecdae694ffc71e7c8ed0fc24e3)
+    ver_str = ver_str.split(' ')[0]
+
+    if version(ver_str) >= version('3.14.8'):
+        return
+    raise RuntimeError(
```

Comment:
Is there any way we can make a single attempt to take corrective action here? 
(Something like, if facter is not installed then install this 3.14,8, else if lower ver is installed then remove it and install this 3.14.8)
Even if this attempt fails then error out with appropriate conclusions?

Reply:
I don't think that it is worth even thinking in that direction.

Adding so much of decision-making at the level of a small tool has a risk to add implicit states. It is not always good when the tool we're using is TOO smart. E.g. the tool was asked to do something, it sees a problem in the input but then it tries to _guess_ what was actually meant, it makes a decision to interpret the input in some other way and then the user/developer comes to the point when (s)he doesn't understand what exactly had happened and why.  Everybody who has ever worked with scripting languages like PHP or Javascript did see the complexity that comes from the thin air just because the language makes too many attempts to guess what was meant (e.g. in PHP the strings that contain numbers may become equal although not all characters match - just because PHP decides that the developer meant that they are numbers (just forgot to cast the string to float, haha), so it compares them as float values and because of IEEE specifics the numbers were decided to be equal enough.

Long story short, I would suggest to keep the things simple and separate responsibilities. Cfgen's responsibility is either generate the configuration or fail if that is not possible. This patch just adds one more pre-validation that can save hours of consequent debugging. Let's not try to save the whole world. It is not always possible.